### PR TITLE
Selecting word occurrences

### PR DESCRIPTION
**Pull request by notdanilo**
### Pull Request Description
This PR enables IntelliJ like word selection by pressing ALT+J.

### Important Notes
- Words are considered to be composed with alphanumeric and underscores.
- ALT+J starts word occurrences selection if the cursor is inside a word or if a word is already selected.

### Checklist
Please include the following checklist in your PR:

- [x] The documentation has been updated if necessary.
- [x] All code conforms to the [Rust](https://github.com/luna/enso/blob/master/doc/rust-style-guide.md), [Scala](https://github.com/luna/enso/blob/master/doc/scala-style-guide.md), [Java](https://github.com/luna/enso/blob/master/doc/java-style-guide.md) or [Haskell](https://github.com/luna/enso/blob/master/doc/haskell-style-guide.md) style guides as appropriate.
- [x] All code has been tested where possible.

### CODE REVIEW

File: lib/core/src/display/shape/text/text_field.rs [hunk 1]
```diff
@@ -122,6 +125,7 @@ shared! { TextField
 
         /// Removes all cursors except one which is set and given point.
         pub fn set_cursor(&mut self, point:Vector2<f32>) {
+            self.word_occurrences = None;
```
Comment (wdanilo) [inline lib/core/src/display/shape/text/text_field.rs]: Minor one - I'm not sure if I would not create method `clear_word_occurences`. But I'm not sure why I have strong feelign it should be a method.

File: lib/core/src/display/shape/text/text_field/word_occurrence.rs [hunk 2]
```diff
@@ -0,0 +1,202 @@
...
+        let chars = content.lines()[line].chars();
+        if let Some(range) = get_index_range_of_word_at(chars,cursor.position.column) {
+            let column = range.start;
+            let start  = TextLocation {line,column};
+            let column = range.end;
+            let end    = TextLocation {line,column};
+            Some(Self::new(start..end))
+        } else {
+            None
+        }
```
Comment (wdanilo) [inline lib/core/src/display/shape/text/text_field/word_occurrence.rs]: map

File: lib/core/src/display/shape/text/text_field/word_occurrence.rs [hunk 3]
```diff
@@ -0,0 +1,202 @@
...
+                    cursor_location.line == current_word.start.line &&
+                        cursor_location.column >= current_word.start.column &&
+                        cursor_location.column <= current_word.end.column
+                }).map(|(index, _)| index).unwrap_or(0);
+                let current_index = current_index - 1;
+                Some(Self { words, current_index }.initialize(&cursor))
+            }
+        } else {
+            None
+        }
```
Comment (wdanilo) [inline lib/core/src/display/shape/text/text_field/word_occurrence.rs]: map

File: lib/core/src/display/shape/text/text_field/word_occurrence.rs [hunk 4]
```diff
@@ -0,0 +1,202 @@
...
+                None
+            } else {
+                let current_index      = words.iter().find_position(|current_word| {
+                    cursor_location.line == current_word.start.line &&
+                        cursor_location.column >= current_word.start.column &&
+                        cursor_location.column <= current_word.end.column
+                }).map(|(index, _)| index).unwrap_or(0);
+                let current_index = current_index - 1;
+                Some(Self { words, current_index }.initialize(&cursor))
+            }
```
Comment (wdanilo) [inline lib/core/src/display/shape/text/text_field/word_occurrence.rs]: use a function for convertong bool to maybe from boolinator (Exported in prelude)

File: lib/core/src/display/shape/text/text_field/word_occurrence.rs [hunk 5]
```diff
@@ -0,0 +1,202 @@
...
+pub struct WordOccurrences {
+    /// Words occurrences.
+    #[shrinkwrap(main_field)]
+    pub words     : Vec<WordRange>,
+    current_index : usize
+}
+
+impl WordOccurrences {
+    /// Gets the range of each occurrence of word.
+    pub fn new(content:&TextFieldContent, cursor:&mut Cursor) -> Option<Self> {
```
Comment (wdanilo) [inline lib/core/src/display/shape/text/text_field/word_occurrence.rs]: This is way too big, needs splitting down.

File: lib/core/src/display/shape/text/text_field/word_occurrence.rs [hunk 6]
```diff
@@ -0,0 +1,202 @@
...
+
+    for word_in_content in get_words(content) {
+        let count = word_in_content.iter().zip(word).filter(|&((_, a), b)| a == b).count();
+        if count == word_in_content.len() {
+            let (start,_) = word_in_content.first().unwrap();
+            let end       = start + word_in_content.len();
+            occurrences.push(*start..end)
+        }
+    }
+
```
Comment (wdanilo) [inline lib/core/src/display/shape/text/text_field/word_occurrence.rs]: -nl

File: lib/ide/src/view/text_editor.rs [hunk 7]
```diff
@@ -101,6 +106,15 @@ impl TextEditor {
...
             }
         });
+
+        let word_occurrence_keys = KeyMask::new_alt_character('j');
+        let text_editor          = Rc::downgrade(&self.rc);
+        keyboard_actions.set_action(word_occurrence_keys,move |_| {
+           if let Some(text_editor) = text_editor.upgrade() {
+               text_editor.borrow_mut().select_next_word_occurrence();
+           }
+        });
```
Comment (farmaazon) [inline lib/ide/src/view/text_editor.rs]: There is set of predefined keyactions for each textfield, and I think selecting word occurrences should be there.
So move this code to lib/core/src/display/shape/text/text_field/frp/keyboard.rs (somewhere at line 154-156)

File: lib/ide/file-manager/mock-server/.gitignore [hunk 8]
```diff
@@ -1 +1 @@
-*.txt
+*.enso
```
Comment (farmaazon) [inline lib/ide/file-manager/mock-server/.gitignore]: :+1:

File: lib/core/src/display/shape/text/text_field/word_occurrence.rs [hunk 9]
```diff
@@ -0,0 +1,202 @@
...
+    }
+}
+
+
+
+// =======================
+// === WordOccurrences ===
+// =======================
+
+/// A struct containing word occurrences in `TextFieldContent`.
```
Comment (farmaazon) [inline lib/core/src/display/shape/text/text_field/word_occurrence.rs]: There should be explanation what is the definition of word. Perfectly if it will be on each structure referring to "words" (and function in TextField which uses this structure), or at least at this module root - but then each structure should refer to this docs (e.g. _for definition of word, see `word_occurrence` module doc_.

File: lib/core/src/display/shape/text/text_field/word_occurrence.rs [hunk 10]
```diff
@@ -0,0 +1,202 @@
...
+mod test {
+    use super::*;
+
+    #[test]
+    fn index_range_of_word_at() {
+        let content           = String::from("   abc    def    ghi  ");
+        let content:Vec<char> = content.chars().collect();
+        let range             = get_index_range_of_word_at(&content,12);
+        assert_eq!(range, Some(10..13));
+    }
```
Comment (farmaazon) [inline lib/core/src/display/shape/text/text_field/word_occurrence.rs]: Please add cases for string begin and end, as well as for empty string. And the _ and numbers should be also covered.

File: lib/core/src/display/shape/text/text_field/word_occurrence.rs [hunk 11]
```diff
@@ -0,0 +1,202 @@
...
+    #[test]
+    fn word_occurrences() {
+        let content = String::from("   abc    def    ghi  abcabc abc");
+        let content:Vec<char> = content.chars().collect();
+        let word              = String::from("abc");
+        let word:Vec<char>    = word.chars().collect();
+        let occurrences       = get_word_occurrences(&content,&word);
+        assert_eq!(occurrences[0],3..6);
+        assert_eq!(occurrences[1],29..32);
+    }
```
Comment (farmaazon) [inline lib/core/src/display/shape/text/text_field/word_occurrence.rs]: We have much logic in WordOccurrences implementation, but we have no tests for it. Please add them.

Mocking TextFieldContent is not easy task, but there are some examples for it, e.g. `get_line_fragment` test. You can make the `mock_properties` function in `content::test` public

File: lib/core/src/display/shape/text/text_field/word_occurrence.rs (hunk 4 above)
Comment (notdanilo) [reply lib/core/src/display/shape/text/text_field/word_occurrence.rs]: I used filter instead :)

File: lib/core/src/display/shape/text/text_field/word_occurrence.rs (hunk 5 above)
[the remark this answers is not in the document]
Comment (notdanilo) [reply lib/core/src/display/shape/text/text_field/word_occurrence.rs]: I improved the doc. Please check if it's clearer now.

File: lib/core/src/display/shape/text/text_field/word_occurrence.rs [hunk 12]
```diff
@@ -180,6 +186,63 @@ fn get_word_occurrences(content:&[char], word:&[char]) -> Vec<Range<usize>> {
...
+        assert_eq!(range, Some(TextLocation {line:0, column:3}..TextLocation{line:0, column:12}));
+
+        let cursor = Cursor::new(TextLocation { line:1, column: 0});
+        let range  = WordOccurrences::get_range(&content,&cursor);
+        assert!(range.is_none());
+
+        let mut cursor = Cursor::new(TextLocation { line:0, column: 4});
+        cursor.select_range(&(TextLocation {line:0, column:3}..TextLocation{line:0, column:12}));
+        let range = WordOccurrences::get_range(&content,&cursor);
+        assert_eq!(range, Some(TextLocation {line:0, column:3}..TextLocation{line:0, column:12}));
```
Comment (farmaazon) [inline lib/core/src/display/shape/text/text_field/word_occurrence.rs]: Add tests for getting range with cursor at end of file, and for getting range in empty content.

File: lib/core/src/display/shape/text/text_field/word_occurrence.rs [hunk 13]
```diff
@@ -180,6 +186,63 @@ fn get_word_occurrences(content:&[char], word:&[char]) -> Vec<Range<usize>> {
...
+        assert!(words.is_none());
+
+        let cursor    = Cursor::new(TextLocation {line:0,column:4});
+        let mut words = WordOccurrences::new(&content,&cursor).expect("Couldn't find words");
+
+        let word1 = WordRange::new(TextLocation{line:0,column:3}..TextLocation{line:0,column:12});
+        let word2 = WordRange::new(TextLocation{line:3,column:2}..TextLocation{line:3,column:11});
+        assert_eq!(words.select_next(), Some(word1));
+        assert_eq!(words.select_next(), Some(word2));
+        assert!(words.select_next().is_none());
```
Comment (farmaazon) [inline lib/core/src/display/shape/text/text_field/word_occurrence.rs]: Add test where WordOccurrences::new would return None. Perhaps we want to test case for empty strings as well.

Comment (notdanilo) [reply lib/core/src/display/shape/text/text_field/word_occurrence.rs]: None is already tested when there is no word occurrence. I have added a test for empty strings.
